sklearn/preprocessing/imputation.py

```diff
@@ -374,3 +375,328 @@ def transform(self, X):
+        """Impute all missing values in X.
+
+        Parameters
+        ----------
+        X : {array-like}, shape = [n_samples, n_features]
+            The input data to complete.
+
+        adjusted_n_neighbors : int
+            Indicates whether to pass n_neighbors or n_neighbors+1 to
+            _tranform().
+            Calling transform() automatically sets this to self.n_neighbors
+            while fit_transform() sets it to self.n_neighbors + 1.
+        """
+        # Import(s) here to avoid circular import
+        from ..neighbors.base import _get_weights
+
+        check_is_fitted(self, 'statistics_')
+        force_all_finite = False if self.missing_values in ["NaN",
+                                                            np.nan] else True
+        X = check_array(X, accept_sparse=False, dtype=FLOAT_DTYPES,
+                        force_all_finite=force_all_finite, copy=self.copy)
+        # Check for +/- inf
+        if np.any(np.isinf(X)):
+            raise ValueError("+/- inf values are not allowed in data to be "
+                             "transformed.")
+
+        # Get fitted data and ensure correct dimension
+        fitted_X = self._fitted_neighbors._fit_X
+        if X.shape[1] != fitted_X.shape[1]:
+            raise ValueError("Incompatible dimension between the fitted "
+                             "dataset and the one to be transformed.")
+        mask = _get_mask(X, self.missing_values)
+        n_rows_X, n_cols_X = X.shape
+        row_total_missing = mask.sum(axis=1)
+        if not np.any(row_total_missing):
+            return X
+
+        # Check for excessive missingness in rows
+        bad_rows = row_total_missing > (mask.shape[1] * self.row_max_missing)
+        if np.any(bad_rows):
+            warnings.warn(
+                "There are rows with more than {0}% missing values. The "
+                "missing features in these rows are imputed with column means."
+                    .format(self.row_max_missing * 100))
+            X_bad = X[bad_rows, :]
+            X = X[~bad_rows, :]
+            mask = mask[~bad_rows]
+            row_total_missing = mask.sum(axis=1)
+        row_has_missing = row_total_missing.astype(np.bool)
+
+        if np.any(row_has_missing):
+            receiver_row_index = np.arange(
+                X.shape[0]).reshape((X.shape[0], 1))[row_has_missing, :]
+            neighbors = self._fitted_neighbors.kneighbors(
+                X[row_has_missing, :], n_neighbors=adjusted_n_neighbors)
+
+            # Get row index, distance, and weights of donors
+            knn_distances, knn_row_index = neighbors
+            # Vertically split sets of k-donor indices
+            knn_row_index = np.vsplit(knn_row_index, knn_row_index.shape[0])
+            row_repeats = row_total_missing[row_total_missing != 0]
+
+            # Weighting: Set self and degenerate donor(s) distance to inf
+            weight_matrix = None
+            if self.weights in ["distance"]:
```

Comment:
This means `weights` is otherwise ignored, despite our saying that we support a callable?

The logic here is hard to follow. I'm not certain how to improve it, except that pulling it into a helper method will help a lot in clarifying its interface and purpose (and perhaps even allowing shorter variable names to be used).

Reply:
This if-block sets self+degenerate-neighbors distance to inf so they contribute zero weight. I was not sure if this would apply to any arbitrary weight so I have not yet added an "...or callable(self.weight)". Do you think it makes sense to allow all callable weights through into the if-block and document the behavior?